runtime/src/main/java/org/corfudb/util/Utils.java

```diff
@@ -119,6 +117,80 @@ public static String toReadableId(UUID id) {
         return Long.toHexString((id.getLeastSignificantBits()) & 0xFFFF);
     }
 
+
+    /**
+     * Verify that the stripes are not empty and that each segment has one stripe.
+     * @param segments segments to validate
+     */
+    private static void validateSegments(List<Layout.LayoutSegment> segments) {
+        checkNotNull(segments);
+        checkArgument(!segments.isEmpty());
+        long previousSegmentEndAddress = 0;
+        for (Layout.LayoutSegment segment : segments) {
+            checkState(segment.getStart() == previousSegmentEndAddress);
+            previousSegmentEndAddress = segment.getEnd();
+            // only supported for chain replication
+            checkArgument(segment.getReplicationMode() == Layout.ReplicationMode.CHAIN_REPLICATION);
+            // Since stripping is not supported, we can assume only one stripe exists
+            checkArgument(segment.getStripes().size() == 1);
+            // A stripe cannot be empty
+            checkArgument(!segment.getStripes().get(0).getLogServers().isEmpty());
+        }
+        // The last segment (i.e open segment) end address is -1 (denoting infinity)
+        checkState(previousSegmentEndAddress == -1);
+    }
+
+    /**
+     * This method selects a head node: the first node (in the open segment) that also exists in all previous
+     * segments.
+     * @param layout layout to use to find the head node
+     * @return head node
+     * @throws IllegalStateException if no such node exists
+     */
+    private static String selectHeadNode(Layout layout) {
+        checkNotNull(layout);
+        validateSegments(layout.getSegments());
+        List<Layout.LayoutSegment> segments = layout.getSegments();
+
+        // select open segment (i.e. last segment)
+        Layout.LayoutSegment openSegment = segments.get(segments.size() - 1);
+        // only support chain replication
+        Layout.LayoutStripe openSegmentStripe = openSegment.getStripes().get(0);
+
+        if (segments.size() == 1) {
+            // Since there is only one segment we can return the first node in the stripe (i.e. head node)
+            return segments.get(0).getStripes().get(0).getLogServers().get(0);
+        }
+
+        // walk the chain of the last segment and find the first node that exists in all previous segments
+        List<Layout.LayoutSegment> reversedSegmentPrefix = segments.subList(0, segments.size() - 1);
+        checkState(!reversedSegmentPrefix.isEmpty());
+        // Reverse the prefix to simplify the indexing when walking the segments backwards (i.e. high to low)
+        Collections.reverse(reversedSegmentPrefix);
```

Comment:
I not understanding why we need to walk the segments backwards? If we need to check all the segments, does it matter if we do forward or backward? If it's for performance consideration, it looks like the first segment has a higher probablility to not having the candidate head node? S1:{n1} S2:{n3, n1} S3:{n3, n1}

Reply:
Correct. Like I mentioned earlier its always possible to compute the sequencer state from only one node, so we need to aggregate. I will simplify this logic by ORing all heads from all segments.